docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,423 @@
+
+This method describes adding a new `spec.asoManagedClusterPatches` field to the existing
+AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+
+	// ASOManagedClusterPatches defines patches to be applied to the generated ASO ManagedCluster resource.
+	ASOManagedClusterPatches []string `json:"asoManagedClusterPatches,omitempty"`
+}
+```
+
+After CAPZ calculates the ASO ManagedCluster for a given AzureManagedControlPlane during a reconciliation, it
+will apply these patches in order to the ManagedCluster before ultimately submitting that to ASO. This allows
+users to specify any AKS feature declaratively within the existing CAPZ spec. The exact format of the patches
+is TBD, but likely one or all of the formats described here:
+https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/
+
+The main drawback of this approach is the fragility of the patches themselves, which can break underneath
+users with a new version of ASO or CAPZ. There is also increased risk that a user could modify the ASO
+resource in a way that breaks CAPZ's ability to reconcile it, though CAPZ could perhaps add some extra
+validation that particularly sensitive fields do not get modified by a set of patches. Compared to other
+options, the syntax for specifying a patch is also more cumbersome than if its equivalent had its own CAPZ API
+field. Given the intent of setting patches is to enable niche use-cases for advanced users, these drawbacks
+may be acceptable.
+
+#### Option 8: Users bring-their-own ASO ManagedCluster resource
+
+CAPZ can already incorporate existing ASO resources into Cluster API Cluster configurations. BYO ASO resources
+are never modified or deleted by CAPZ directly but are still read and play in to the status of CAPZ resources.
+
+The flow for this method is roughly:
+1. User creates an ASO ManagedCluster before or at the same time as the other Cluster API and CAPZ resources.
+1. Any updates to the AKS cluster are done by the user through the ASO resource.
+1. The user may control whether or not the ASO ManagedCluster gets deleted with the rest of the Cluster API
+   Cluster by choosing whether or not to pre-create the ASO ResourceGroup.
+
+Benefits of this approach:
+- Minimal or zero additional cost for CAPZ to implement outside of a new test. Users can try this today with
+  the latest version of CAPZ.
+- Freedom for users to craft the ASO ManagedCluster and other resources exactly how they like, even a
+  different API version (including preview).
+- Protection from CAPZ being responsible for managing AKS configuration that it does not understand.
+
+Drawbacks:
+- Requires users to rework templates to add ASO resources, but only if they want AKS features not exposed by
+  CAPZ.
+- Doesn't support ClusterClass.
+- CAPZ may no longer be able to fulfill CAPI's contract by modifying fields like Kubernetes version or replica
```

Comment:
How much discovery work to be able to more accurately scope the concrete changes needed to CAPZ that would support this option? Seems like this option requires very little work (supporting documentation would probably be a larger effort than any new code + tests).

Reply:
Not much I don't think. I'd like to actually try this for myself to make sure there are no obvious blockers that I missed here. I could do that within a day.

Specifically for this point about the CAPI contract, that would maybe take an extra day for me to confidently know whether or not that's possible to overcome.